File: Ballast.Core/src/Client/messaging/events/sign-in/player-signed-out-event.ts
import { IPlayer } from '../../../models/player';
import { IEvent } from '../../event';
import { EventBase } from '../../event-base';

export interface IPlayerSignedOutEvent extends IEvent {
    readonly player?: IPlayer;
}

export class PlayerSignedOutEvent extends EventBase implements IPlayerSignedOutEvent {

    public static readonly id: string = 'PlayerSignedOutEvent';

    public get id() {
        return PlayerSignedOutEvent.id;
    }

    public readonly player?: IPlayer; 

    public constructor(player?: IPlayer)
    public constructor(player?: IPlayer, isoDateTime?: string) {
        super(isoDateTime);
        this.player = player;
    }

}

File: ballast-core/src/dotnet/Messaging/Events/PlayerSignedOutEvent.cs
using Ballast.Core.Models;
using System;

namespace Ballast.Core.Messaging.Events
{

    public class PlayerSignedOutEventState : EventStateBase
    {
        public Player Player { get; set; }
    }

    public class PlayerSignedOutEvent : EventBase 
    {

        public override string Id => nameof(PlayerSignedOutEvent);

        public Player Player { get; private set; } 

        private PlayerSignedOutEvent(Player player = null, string isoDateTime = null) : base(isoDateTime) {
            Player = player;
        }

        public static implicit operator PlayerSignedOutEvent(PlayerSignedOutEventState state) =>
            new PlayerSignedOutEvent(state.Player, state.IsoDateTime);

        public static PlayerSignedOutEvent FromPlayer(Player player = null) =>
            new PlayerSignedOutEvent(player);

            
    }

}